### PR TITLE
GDoc tagging

### PR DESCRIPTION
**Pull request by ikesau**
Adds a table and UI to use the tag system for google docs

Also ports over tags from WP posts when using the "Create Gdoc" feature.

Using the `ManyToMany` feature of TypeORM to create entries in the join table. I was getting [this issue](https://github.com/typeorm/typeorm/issues/8533) but after clearing my build directory it went away. Will need to do some more testing to see if I can work out what was causing it.

### CODE REVIEW

File: db/migration/1680210665372-PostsGdocsXTags.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
+
+export class PostsGdocsXTags1680210665372 implements MigrationInterface {
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        queryRunner.query(`
+        CREATE TABLE posts_gdocs_x_tags (
+            gdocId VARCHAR(255) NOT NULL,
+            tagId INT NOT NULL,
+            PRIMARY KEY (gdocId,tagId),
+            KEY FK_post_tags_tag_id (tagId),
```
Comment (marcelgerber) [inline db/migration/1680210665372-PostsGdocsXTags.ts:10]: don't think you need that line?

**Review by mlbrgl — COMMENTED** [review]
hey Ike, I'm trying something new: going through PR reviews with a predefined and visible checklist. This helps me (the reviewer) be more consistent, and you (the submitter) hopefully get a better sense of progress. 

- [x] is master merged?
- [x] what problem is being solved here?
we want to:		
  - surface all the content related to a topic (all charts related to a topic in exhaustive /charts page and all charts blocks, filter in search, homepage grid, possibly in admin tools)
  - contextualize an article or a subtopic (citation inheritance from topic?) and provide navigational aids to make sense of and navigate the hierarchy (breadcrumb, back to parent topic button)
- [x] what solution has been chosen?
	- attaching content to the existing hierarchical grapher tag structure
- [ ] are there other solutions?
	- needing to surface a hierarchy of topics and query against it doesn't necessarily mean we need to rely on an additional topic tree separate from the topics themselves. We could imagine creating a tree from organizing the topic pages between themselves, rather than maintaining a tree of tags and attaching topic pages to them. 

_@ikesau I know we touched on this a few times, but I'm still not fully confident in the rationale for an additional data structure here. You've also done [some deeper thinking on this](https://www.notion.so/owid/2023-03-28-Categories-in-Google-Docs-bd094b92818d4614be79540b0a9fadb6#1f9dfc726b4f42d08cde26bda70ccc41) so I'd be interested to hear your take before going further with the review._

- [ ] is the scope small, clear and well-defined?
- [ ] could a flow chart / sequence diagram / architecture diagram help?
- [ ] what is the happy path?
- [ ] could the code be easier to read and understand?
- [ ] can I help increase our confidence in the implementation by contributing tests?
- [ ] should some errors be made more prominent?
- [ ] what is missing, what should have been added / done / changed but hasn't?

**Comment by ikesau** [comment]
To make sure I understand your suggestion, you're saying we could make categorization work by tracking the links each topic page has to one another? 

I hadn't really thought about that as an alternative. Creating a graph DB structure in MySQL seems more convoluted to implement and more awkward to manage alongside our Grapher tagging system, with little obvious benefit.

Plus if we did it that way, we'd also need to merge the two systems for as long as we still had WP posts which are tagged using the Grapher system

What are the benefits you can see behind a graph system?

**Comment by mlbrgl** [comment]
The way I see it is that we are aiming at merging grapher tags and topic pages (from a categorization point of view). There shouldn't be grapher tags representing topics outside of those for which we have topic pages. So ultimately, grapher tags and topic pages represent the same thing.

Bijection

_from https://en.wikipedia.org/wiki/Bijection_

By maintaining grapher tags, we're keeping around a data structure that needs to be kept in sync with topics, when we could avoid it. And whether we create a DB relation between a `posts_gdocs` and a `tags` or two `posts_gdocs` doesn't seem all that different to me.

Just to be clear, I'm not suggesting to build this hierarchy out of the links in the content of each topic page. These parent / child relationships would be identified as such in a designated field or block.

You're right though, the transition would need some more thinking. And maybe going through this solution is a necessary step. Ordering would need to be considered too, but so does grapher tags ordering.

**Comment by ikesau** [comment]
As far as I can tell the reason the posts and charts got out of sync was because there was no glue that attached WP categories to their grapher equivalents, and nothing that enforced that mapping.

With this new setup, everything is using the same tagging structure so there's no way for a new category to arise that isn't part of the system both entities are using.

If the ETL changes things, we can cross that bridge soon, but I don't think we should block this PR until then unless you think there's a large downside to this strategy or a clear upside to rewriting everything to be Graph-based.

File: adminSiteServer/apiRouter.ts [hunk 2]
```diff
@@ -2318,10 +2322,12 @@ apiRouter.post("/posts/:postId/createGdoc", async (req: Request) => {
 
         const gdoc = new Gdoc(gdocId)
         gdoc.slug = post.slug
+        gdoc.tags = tags
+        gdoc.content.title = post.title
         gdoc.published = false
         gdoc.createdAt = new Date()
         gdoc.publishedAt = post.published_at
-        await dataSource.getRepository(Gdoc).insert(gdoc)
+        await dataSource.getRepository(Gdoc).save(gdoc)
```
Comment (mlbrgl) [inline adminSiteServer/apiRouter.ts:2330]: if I get this right, `insert` doesn't cascade the tags relationship, while `save` does. Maybe worth adding a note about this here?

File: baker/algolia/indexToAlgolia.ts [hunk 3]
```diff
@@ -141,12 +141,13 @@ function generateGdocRecords(
...
     const records: PageRecord[] = []
     for (const gdoc of gdocs) {
+        if (!gdoc.content.body) continue
         // Only rendering the blocks - not the page nav, title, byline, etc
         const renderedPostContent = ReactDOMServer.renderToStaticMarkup(
             React.createElement(
                 "div",
                 {},
-                gdoc.content.body?.map((block) => ArticleBlock({ b: block }))
+                ArticleBlocks({ blocks: gdoc.content.body })
```
Comment (mlbrgl) [inline baker/algolia/indexToAlgolia.ts:150]: nit: shouldn't we just make the file a .tsx and make this clearer?

File: db/model/GdocXTag.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { Entity, BaseEntity, PrimaryColumn } from "typeorm"
+
+@Entity("posts_gdocs_x_tags")
+export class GdocXTag extends BaseEntity {
+    static table = "posts_gdocs_x_tags"
```
Comment (mlbrgl) [inline db/model/GdocXTag.ts:5]: I don't know what this is for

File: db/model/Gdoc/Gdoc.ts [hunk 5]
```diff
@@ -37,6 +40,25 @@ import { archieToEnriched } from "./archieToEnriched.js"
 import { Link } from "../Link.js"
 import { imageStore } from "../Image.js"
 
+@Entity("tags")
+export class Tag extends BaseEntity implements OwidGdocTag {
+    static table = "tags"
```
Comment (mlbrgl) [inline db/model/Gdoc/Gdoc.ts:45]: I don't know what this is for

**Review by mlbrgl — APPROVED** [review]
> With this new setup, everything is using the same tagging structure so there's no way for a new category to arise that isn't part of the system both entities are using.

True. I was actually talking about topic pages and tags getting out of sync. If we rename the "Poverty" topic page, we need to rename the "Poverty" tag in parallel since they are different things. Similarly for the other CRUD operations on tags or topic pages.

This is a pain authors might experience more at the beginning while the migration is being carried out, probably less so as the tag structure settles. 

A couple of PR disclaimers:
- I'm not too sure about how TypeORM expresses many to many relationships, so my approval on this front is not a strong signal
- the Algolia export looks good but I haven't run it. I can spend more time on this on Monday if neceessary.

File: db/model/Gdoc/Gdoc.ts [hunk 6]
```diff
@@ -37,6 +40,25 @@ import { archieToEnriched } from "./archieToEnriched.js"
...
+    @Column({ nullable: true }) updatedAt!: Date
+    @Column({ nullable: true }) parentId!: number
+    @Column() isBulkImport!: boolean
+    @Column() specialType!: string
+    @ManyToMany(() => Gdoc, (gdoc) => gdoc.tags)
+    @JoinTable({
+        name: "posts_gdocs_x_tags",
+        joinColumn: { name: "tagId", referencedColumnName: "id" },
+        inverseJoinColumn: { name: "gdocId", referencedColumnName: "id" },
+    })
```
[the remark this answers is not in the document]
Comment (ikesau) [reply db/model/Gdoc/Gdoc.ts]: Oh, good catch. Looks like everything works without it. 👍

File: db/model/GdocXTag.ts (hunk 4 above)
Comment (ikesau) [reply db/model/GdocXTag.ts:5]: It appears to be a convention to use when making a Knex query (e.g. `await db.knexTable(Gdoc.table)`) so I figured I'd just put it everywhere

**Pull request merged**